controllers/librarypanel_controller.go

```diff
@@ -0,0 +1,318 @@
+	for _, grafana := range instances {
+		err := r.reconcileWithInstance(ctx, &grafana, libraryPanel, contentModel, hash)
+		if err != nil {
+			applyErrors[fmt.Sprintf("%s/%s", grafana.Namespace, grafana.Name)] = err.Error()
+		}
+	}
+
+	condition := buildSynchronizedCondition("Library panel", conditionLibraryPanelSynchronized, libraryPanel.Generation, applyErrors, len(instances))
+	meta.SetStatusCondition(&libraryPanel.Status.Conditions, condition)
+
+	contentUID := fmt.Sprintf("%s", contentModel["uid"])
+	libraryPanel.Status.Hash = hash
+	libraryPanel.Status.UID = content.CustomUIDOrUID(libraryPanel, contentUID)
+	if err := r.Client.Status().Update(ctx, libraryPanel); err != nil {
+		log.Error(err, "failed to update status")
+	}
+
+	if len(applyErrors) > 0 {
+		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
+	}
+
+	return ctrl.Result{RequeueAfter: libraryPanel.Spec.ResyncPeriod.Duration}, nil
+}
+
+func (r *GrafanaLibraryPanelReconciler) reconcileWithInstance(ctx context.Context, instance *v1beta1.Grafana, cr *v1beta1.GrafanaLibraryPanel, model map[string]any, hash string) error {
+	if instance.IsInternal() {
+		err := ReconcilePlugins(ctx, r.Client, r.Scheme, instance, cr.Spec.Plugins, fmt.Sprintf("%v-librarypanel", cr.Name))
+		if err != nil {
+			return err
+		}
+	} else if instance.IsExternal() && cr.Spec.Plugins != nil {
+		return fmt.Errorf("external grafana instances don't support plugins, please remove spec.plugins from your library panel cr")
+	}
+
+	grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return err
+	}
+
+	folderUID, err := getFolderUID(ctx, r.Client, cr)
+	if err != nil {
+		return err
+	}
+
+	uid := content.CustomUIDOrUID(cr, fmt.Sprintf("%s", model["uid"]))
+	name := fmt.Sprintf("%s", model["name"])
+
+	defer func() {
+		instance.Status.LibraryPanels = instance.Status.LibraryPanels.Add(cr.Namespace, cr.Name, uid)
+		r.Client.Status().Update(ctx, instance)
+	}()
+
+	resp, err := grafanaClient.LibraryElements.GetLibraryElementByUID(uid)
+
+	var panelNotFound *library_elements.GetLibraryElementByUIDNotFound
+	if err != nil {
+		if !errors.As(err, &panelNotFound) {
+			return err
+		}
+
+		// doesn't yet exist--should provision
+		// nolint:errcheck
+		if _, err = grafanaClient.LibraryElements.CreateLibraryElement(&models.CreateLibraryElementCommand{
+			FolderUID: folderUID,
+			Kind:      int64(libraryElementTypePanel),
+			Model:     model,
+			Name:      name,
+			UID:       uid,
+		}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// it can happen that the user does not utilize `.spec.uid` but updates
+	// the UID within the content model itself. this will create a conflict b/c
+	// we are effectively requesting a change to the uid, which is immutable.
+	if content.IsUpdatedUID(cr, uid) {
+		return fmt.Errorf("library panel uid is immutable, but was updated on the content model")
+	}
```

Comment:
Since modifying the 'inner uid' in the model is explicitly disabled here, I would move the check out of `reconcileWithInstance` and instead have it as a pre-condition before the apply loop.

When there is an error we then add an `InvalidSpec` condition telling the users why the reconcile failed on the resource.
This is important to users that have one team responsible for the operator but other teams that can apply resources in their own namespaces and have them synchronized to Grafana instances otherwise out of their control.

Additionally, the error will only show up once rather than multiplied by the [number of matching instances](https://github.com/grafana/grafana-operator/blob/master/controllers/controller_shared.go#L309)